torch/_inductor/codegen/simd.py

```diff
@@ -1701,85 +1995,113 @@ def select_tiling(cls, node_schedule, numel, reduction_numel=sympy.Integer(1)):
-                    b0, b1 = ranked_tilings[0]
+                    (a0, a1), (b0, b1) = (b0, b1), (a0, a1)
                 assert V.graph.sizevars.size_hint(a1 - b1) > 0
-                if V.graph.sizevars.statically_known_multiple_of(a1, b1):
-                    tiling = (a0, FloorDiv(a1, b1), b1)
-                    ranked_tilings = [tiling] + ranked_tilings
+                if not V.graph.sizevars.statically_known_multiple_of(a1, b1):
+                    return None
+
+                new_tiling = {
+                    "z": a0,
+                    "y": FloorDiv(a1, b1),
+                    "x": b1,
+                    "r0_": reduction_numel,
+                }
+
+                # Copy the existing reduction tiling
+                new_tiling.update(
+                    {
+                        dim: split
+                        for dim, split in tiling0.items()
+                        if dim not in new_tiling
+                    }
+                )
+
+                return new_tiling
+
+            for i in range(1, len(ranked_tilings)):
+                new_3d_tiling = convert_tiling_to_3d(
+                    ranked_tilings[0], ranked_tilings[i]
+                )
+                if new_3d_tiling is not None:
+                    ranked_tilings = [new_3d_tiling] + ranked_tilings
                     break  # only 1 choice for now
 
         if len(ranked_tilings) > 1:
             perf_hint_log.info("possibly bad tiling: %s", ranked_tilings)
 
         # Optionally, prefer tiling into as many dimensions as possible.
         if config.triton.prefer_nd_tiling:
-            # Get candidate tilings from the node ranges.
-            node_ranges = [
-                node.get_ranges()[0]
-                for node in EnableReduction.filter(node_schedule)
-                if isinstance(node, scheduler.SchedulerNode)
-            ]
-            new_tilings: OrderedSet[Tuple[sympy.Expr]] = OrderedSet()
-            for node_range in node_ranges:
-                # Collapse leading dims, to fit in the maximum dimensionality.
-                num_leading_dims = max(0, len(node_range) - config.triton.max_tiles)
-                first_trailing_dim = num_leading_dims + 1
-                collapsed_leading_dim = sympy_product(node_range[:first_trailing_dim])
-                tiling = [collapsed_leading_dim] + list(node_range[first_trailing_dim:])
-                new_tilings.add(tuple(tiling))
+            ranked_tilings = (
+                cls.get_nd_tilings(node_schedule, is_pointwise) + ranked_tilings
+            )
 
-            # Rank tilings by the number of dimensions. E.g., prefer 2D to 1D.
-            # Since this is a stable sort, ties are broken by schedule order.
-            ranked_new_tilings = sorted(new_tilings, key=len, reverse=True)
-            ranked_tilings = ranked_new_tilings + ranked_tilings
+        def is_compatible_with_numels(tiling: Dict[str, sympy.Expr]) -> bool:
```

Comment:
This is a good sanity check, but I'm not sure we need it anymore. I'll try removing it.

Reply:
Interesting, it seems like it's needed for these two tests:
```
 test/inductor/test_torchinductor.py::GPUTests::test_softmax_one_kernel_loop_cuda
  test/inductor/test_torchinductor.py::GPUTests::test_softmax_one_kernel_persist_cuda
```

I'll look into why.

Reply:
Looks like we don't need this anymore. I'm not sure what changed, but tests seem to pass without it now. Reverted in [76ed913](https://github.com/pytorch/pytorch/pull/137243/commits/76ed913328dbeef8168a9eff023f6e4d1e5b5002).